File: ml/features/extractor.py
"""
Cyber Chaukidaar - Python Seismic Vibration Feature Extractor
Extracts 8 critical features matching the JS extractor:
1. RMS
2. Peak
3. Peak-to-Peak
4. Variance
5. Dominant Frequency
6. Spectral Energy
7. Spectral Centroid
8. Inter-Peak Interval
"""

import numpy as np

class FeatureExtractor:
    def __init__(self, sampling_rate=100, peak_threshold_ratio=0.4):
        self.sampling_rate = sampling_rate
        self.peak_threshold_ratio = peak_threshold_ratio

    def compute_magnitudes(self, samples):
        arr = np.array(samples)
        if arr.ndim == 2 and arr.shape[1] >= 3:
            return np.sqrt(arr[:, 0]**2 + arr[:, 1]**2 + arr[:, 2]**2)
        return arr.flatten()

    def remove_dc_offset(self, signal):
        return signal - np.mean(signal)

    def extract(self, samples):
        raw = self.compute_magnitudes(samples)
        if len(raw) < 4:
            return self.get_default_features()

        signal = self.remove_dc_offset(raw)
        N = len(signal)

        # 1. RMS
        rms = float(np.sqrt(np.mean(signal**2)))

        # 2. Peak & 3. Peak-to-Peak
        peak = float(np.max(np.abs(signal)))
        peak_to_peak = float(np.max(signal) - np.min(signal))

        # 4. Variance
        variance = float(np.var(signal))

        # 5, 6, 7. Spectral features
        fft_vals = np.fft.rfft(signal)
        fft_mags = np.abs(fft_vals) / N
        freqs = np.fft.rfftfreq(N, 1.0 / self.sampling_rate)

        # Skip DC component at index 0
        mags_no_dc = fft_mags[1:]
        freqs_no_dc = freqs[1:]

        if len(mags_no_dc) > 0 and np.sum(mags_no_dc) > 0:
            dom_idx = np.argmax(mags_no_dc)
            dom_freq = float(freqs_no_dc[dom_idx])
            spec_energy = float(np.sum(mags_no_dc**2))
            spec_centroid = float(np.sum(freqs_no_dc * mags_no_dc) / np.sum(mags_no_dc))
        else:
            dom_freq = 0.0
            spec_energy = 0.0
            spec_centroid = 0.0

        # 8. Inter-Peak Interval
        ipi = self.compute_ipi(signal, peak, self.sampling_rate)

        return {
            "rms": round(rms, 4),
            "peak": round(peak, 4),
            "peakToPeak": round(peak_to_peak, 4),
            "variance": round(variance, 6),
            "dominantFrequency": round(dom_freq, 2),
            "spectralEnergy": round(spec_energy, 4),
            "spectralCentroid": round(spec_centroid, 2),
            "interPeakInterval": round(ipi, 2)
        }

    def compute_ipi(self, signal, max_abs, Fs):
        thresh = max_abs * self.peak_threshold_ratio
        peaks = []
        for i in range(1, len(signal) - 1):
            if signal[i] > thresh and signal[i] > signal[i-1] and signal[i] >= signal[i+1]:
                if len(peaks) == 0 or (i - peaks[-1]) >= 5:
                    peaks.append(i)
        if len(peaks) < 2:
            return 0.0
        intervals_ms = [(peaks[j] - peaks[j-1]) / Fs * 1000.0 for j in range(1, len(peaks))]
        return float(np.mean(intervals_ms))

    def get_default_features(self):
        return {
            "rms": 0.0, "peak": 0.0, "peakToPeak": 0.0, "variance": 0.0,
            "dominantFrequency": 0.0, "spectralEnergy": 0.0, "spectralCentroid": 0.0,
            "interPeakInterval": 0.0
        }

File: ml/features/extractor.js
/**
 * Cyber Chaukidaar - Seismic Vibration Feature Extractor
 * Extracts 8 critical time & frequency domain features from raw accelerometer signals:
 * 1. RMS (Root Mean Square)
 * 2. Peak (Maximum absolute amplitude)
 * 3. Peak-to-Peak (Span between max and min)
 * 4. Variance (Statistical spread / dispersion)
 * 5. Dominant Frequency (Hz of peak spectral magnitude)
 * 6. Spectral Energy (Sum of squared FFT magnitudes)
 * 7. Spectral Centroid (Center of gravity of frequency spectrum)
 * 8. Inter-Peak Interval (Average ms between prominent local peaks)
 */

class FeatureExtractor {
  /**
   * @param {Object} options
   * @param {number} options.samplingRate - Accelerometer sampling frequency in Hz (e.g., 100 Hz)
   * @param {number} options.peakThresholdRatio - Threshold multiplier for peak detection in IPI
   */
  constructor(options = {}) {
    this.samplingRate = options.samplingRate || 100; // default 100Hz for ADXL345
    this.peakThresholdRatio = options.peakThresholdRatio || 0.4;
  }

  /**
   * Compute magnitude array from 3-axis readings or 1D stream
   * magnitude = sqrt(x² + y² + z²)
   */
  computeMagnitudes(samples) {
    if (!Array.isArray(samples) || samples.length === 0) return [];
    
    // If samples are already numbers
    if (typeof samples[0] === 'number') {
      return samples;
    }

    // If samples are objects {x, y, z}
    return samples.map(s => {
      const x = s.x || 0;
      const y = s.y || 0;
      const z = s.z || 0;
      return Math.sqrt(x * x + y * y + z * z);
    });
  }

  /**
   * Remove DC offset / baseline gravity component (High-pass filter approximation)
   */
  removeDcOffset(signal) {
    if (signal.length === 0) return [];
    const mean = signal.reduce((a, b) => a + b, 0) / signal.length;
    return signal.map(v => v - mean);
  }

  /**
   * Extract all 8 features from a signal window
   * @param {number[]|Object[]} samples - Array of acceleration samples
   * @returns {Object} 8 extracted features
   */
  extract(samples) {
    const rawMagnitudes = this.computeMagnitudes(samples);
    if (rawMagnitudes.length === 0) {
      return this.getDefaultFeatures();
    }

    // High-pass filter / remove DC gravity component
    const signal = this.removeDcOffset(rawMagnitudes);
    const N = signal.length;
    if (N < 4) return this.getDefaultFeatures();

    // 1. RMS
    const sumSquares = signal.reduce((acc, v) => acc + v * v, 0);
    const rms = Math.sqrt(sumSquares / N);

    // 2. Peak & 3. Peak-to-Peak
    let minVal = signal[0];
    let maxVal = signal[0];
    let maxAbs = Math.abs(signal[0]);
    for (let i = 1; i < N; i++) {
      const v = signal[i];
      const absV = Math.abs(v);
      if (v < minVal) minVal = v;
      if (v > maxVal) maxVal = v;
      if (absV > maxAbs) maxAbs = absV;
    }
    const peak = maxAbs;
    const peakToPeak = maxVal - minVal;

    // 4. Variance
    const mean = signal.reduce((acc, v) => acc + v, 0) / N;
    const variance = signal.reduce((acc, v) => acc + Math.pow(v - mean, 2), 0) / N;

    // 5, 6, 7. Frequency Domain via Real Discrete Fourier Transform (DFT)
    const { dominantFrequency, spectralEnergy, spectralCentroid } = this.computeSpectralFeatures(signal, this.samplingRate);

    // 8. Inter-Peak Interval (IPI)
    const interPeakInterval = this.computeInterPeakInterval(signal, maxAbs, this.samplingRate);

    return {
      rms: Number(rms.toFixed(4)),
      peak: Number(peak.toFixed(4)),
      peakToPeak: Number(peakToPeak.toFixed(4)),
      variance: Number(variance.toFixed(6)),
      dominantFrequency: Number(dominantFrequency.toFixed(2)),
      spectralEnergy: Number(spectralEnergy.toFixed(4)),
      spectralCentroid: Number(spectralCentroid.toFixed(2)),
      interPeakInterval: Number(interPeakInterval.toFixed(2))
    };
  }

  /**
   * Spectral feature computation (DFT for frequency domain)
   */
  computeSpectralFeatures(signal, Fs) {
    const N = signal.length;
    const halfN = Math.floor(N / 2);
    let maxMag = -1;
    let domFreq = 0;
    let totalEnergy = 0;
    let weightedFreqSum = 0;
    let magSum = 0;

    // Compute DFT up to Nyquist frequency
    for (let k = 1; k < halfN; k++) {
      let real = 0;
      let imag = 0;
      const angle = (2 * Math.PI * k) / N;
      for (let n = 0; n < N; n++) {
        real += signal[n] * Math.cos(angle * n);
        imag -= signal[n] * Math.sin(angle * n);
      }
      const mag = Math.sqrt(real * real + imag * imag) / N;
      const freq = (k * Fs) / N;
      const energy = mag * mag;

      totalEnergy += energy;
      weightedFreqSum += freq * mag;
      magSum += mag;

      if (mag > maxMag) {
        maxMag = mag;
        domFreq = freq;
      }
    }

    const spectralCentroid = magSum > 0 ? weightedFreqSum / magSum : 0;

    return {
      dominantFrequency: domFreq,
      spectralEnergy: totalEnergy,
      spectralCentroid: spectralCentroid
    };
  }

  /**
   * Inter-Peak Interval (ms between prominent local shockwave peaks)
   */
  computeInterPeakInterval(signal, maxAbs, Fs) {
    const N = signal.length;
    const threshold = maxAbs * this.peakThresholdRatio;
    const peakIndices = [];

    for (let i = 1; i < N - 1; i++) {
      if (signal[i] > threshold && signal[i] > signal[i - 1] && signal[i] >= signal[i + 1]) {
        // Enforce minimum refractory period (e.g. 5 samples = 50ms)
        if (peakIndices.length === 0 || (i - peakIndices[peakIndices.length - 1]) >= 5) {
          peakIndices.push(i);
        }
      }
    }

    if (peakIndices.length < 2) {
      return 0; // Not enough peaks to establish rhythm
    }

    let totalIntervalMs = 0;
    for (let j = 1; j < peakIndices.length; j++) {
      const sampleDiff = peakIndices[j] - peakIndices[j - 1];
      const intervalMs = (sampleDiff / Fs) * 1000;
      totalIntervalMs += intervalMs;
    }

    return totalIntervalMs / (peakIndices.length - 1);
  }

  getDefaultFeatures() {
    return {
      rms: 0,
      peak: 0,
      peakToPeak: 0,
      variance: 0,
      dominantFrequency: 0,
      spectralEnergy: 0,
      spectralCentroid: 0,
      interPeakInterval: 0
    };
  }
}

module.exports = FeatureExtractor;
